builder/dockerfile/parser/parser.go

```diff
@@ -107,29 +108,31 @@ func (d *Directive) setEscapeToken(s string) error {
 	return nil
 }
 
-// processLine looks for a parser directive '# escapeToken=<char>. Parser
-// directives must precede any builder instruction or other comments, and cannot
-// be repeated.
-func (d *Directive) processLine(line string) error {
+// possibleParserDirective looks for one or more parser directives '# escapeToken=<char>' and
+// '# platform=<string>'. Parser directives must precede any builder instruction
+// or other comments, and cannot be repeated.
+func (d *Directive) possibleParserDirective(line string) error {
 	if d.processingComplete {
 		return nil
 	}
-	// Processing is finished after the first call
-	defer func() { d.processingComplete = true }()
 
+	logrus.Debugf("[BUILDER] considering parser directive in %q", line)
 	tecMatch := tokenEscapeCommand.FindStringSubmatch(strings.ToLower(line))
-	if len(tecMatch) == 0 {
-		return nil
-	}
-	if d.escapeSeen == true {
-		return errors.New("only one escape parser directive can be used")
-	}
-	for i, n := range tokenEscapeCommand.SubexpNames() {
-		if n == "escapechar" {
-			d.escapeSeen = true
-			return d.setEscapeToken(tecMatch[i])
+	if len(tecMatch) != 0 {
+		if d.escapeSeen == true {
+			return errors.New("only one escape parser directive can be used")
+		}
+		for i, n := range tokenEscapeCommand.SubexpNames() {
+			if n == "escapechar" {
+				logrus.Debugf("[BUILDER] escape set to '%s'", tecMatch[i])
+				d.escapeSeen = true
+				return d.setEscapeToken(tecMatch[i])
+			}
 		}
 	}
+
+	logrus.Debugf("[BUILDER] no more parser directives allowed from this point")
+	d.processingComplete = true
```

Comment:
I'm not sure that's the right thing. There's really no condition here - if it hits the end of the function without picking up a directive, it really is "game over" in terms of looking for more parser directives.

I was thinking that adding a second directive would be something like this by just adding a second code block mirroring the existing one for the escape token.


```golang
func (d *Directive) possibleParserDirective(line string) error {
	if d.processingComplete {
		return nil
	}

	logrus.Debugf("[BUILDER] considering parser directive in %q", line)
	tecMatch := tokenEscapeCommand.FindStringSubmatch(strings.ToLower(line))
	if len(tecMatch) != 0 {
		for i, n := range tokenEscapeCommand.SubexpNames() {
			if n == "escapechar" {
				if d.escapeSeen == true {
					return errors.New("only one escape parser directive can be used")
				}
				logrus.Debugf("[BUILDER] escape set to '%s'", tecMatch[i])
				d.escapeSeen = true
				return d.setEscapeToken(tecMatch[i])
			}
		}
	}

	tec2Match := tokenSomeOtherParserDirective.FindStringSubmatch(strings.ToLower(line))
	if len(tec2Match) != 0 {
		for i, n := range tokenSomeOtherParserDirective.SubexpNames() {
			if n == "someotherparserdirective" {
				if d.someotherparserdirective == true {
					return errors.New("only one someotherparserdirective can be used")
				}
				logrus.Debugf("[BUILDER] someotherparserdirective set to '%s'", tec2Match[i])
				d.someotherparserdirective = true
				return d.setSomeOtherParserDirective(tec2Match[i])
			}
		}
	}

	logrus.Debugf("[BUILDER] no more parser directives allowed from this point")
	d.processingComplete = true
	return nil
}
```

I wasn't sure what you meant by removing the `defer func()` - there aren't any `defer` statements in this file....